docs/Consuming.md

```diff
@@ -95,6 +95,25 @@ consumer.run({
 
 In the example above, if the consumer is shutting down in the middle of the batch, the remaining messages won't be resolved and therefore not committed. This way, you can quickly shut down the consumer without losing/skipping any messages.
 
+## <a name="concurrent-processing"></a> Partition-aware concurrency
+
+By default, [`eachMessage`](Consuming.md#each-message) is invoked sequentially for each message in each partition. In order to concurrently process several messages per once, you can increase the `partitionsConsumedConcurrently` option:
+
+```javascript
+consumer.run({
+    partitionsConsumedConcurrently: 3, // Default: 1
+    eachMessage: async ({ topic, partition, message }) => {
+        // This will be called up to 3 times concurrently
+    },
+})
+```
+
+Messages in the same partition are still guaranteed to be processed in order, but messages from multiple partitions can be processed at the same time. If `eachMessage` consists of asynchronous work, such as network requests or other I/O, this can improve performance. If `eachMessage` is entirely synchronous, this will make no difference.
+
+The same thing applies if you are using [`eachBatch`](Consuming.md#each-batch). Given `partitionsConsumedConcurrently > 1`, you will be able to process multiple batches concurrently.
+
+A guideline for setting `partitionsConsumedConcurrently` would be that it should not be larger than the number of partitions consumed. Depending on whether or not your workload is CPU bound, it may also not benefit you to set it to a higher number than the number of logical CPU cores. A recommendation is to start with a low number and measure if increasing leads to higher throughput.
```

Comment:
Is the number of CPU cores relevant for a Node application?  This is still single threaded processing.

Reply:
It depends on your workload. If the message processing is entirely synchronous, then this setting is completely irrelevant as all the work takes place on the main thread and will thus be executed sequentially regardless.

However, in most cases the work is not entirely synchronous. In many cases, we call remote APIs or read/write to a database, etc. In those cases, some of the work takes place outside of the main thread, such as when doing network I/O. If each message requires you to make a network call to some API, you would normally just sit around and wait while the remote service is doing some work, but with this concurrency model you can instead fire off multiple of these requests concurrently.

So then you might say that you would just set the concurrency limit to whatever number of partitions you have, in order to achieve maximum concurrency, but that may not actually increase throughput for multiple reasons. First, your downstream dependency might not appreciate getting that many concurrent requests and it's likely their performance may suffer. Even if that's fine, network or disk I/O is not infinitely parallelisable, and correlates with the number of cores you have.

While they're talking about connection pool sizing in this article, the principles are very much the same, and it's a really good high-level explanation: https://github.com/brettwooldridge/HikariCP/wiki/About-Pool-Sizing

So the complex answer is that it really depends on your workload and your dependencies. The recommendation is pretty conservative, because otherwise people will put a gigantic number in there thinking that "more is better" and end up getting lower throughput and potentially really bursty behavior. Throughput may very well increase with higher concurrency, given the right conditions, but it's better to start low and increase it over time while measuring the effect until you hit the sweet spot.